common/persistence/sql/sqlplugin/auth/rds_auth_plugin.go

```diff
@@ -0,0 +1,142 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	AWSConfig "github.com/aws/aws-sdk-go-v2/config"
+	AWSAuth "github.com/aws/aws-sdk-go-v2/feature/rds/auth"
+
+	"go.temporal.io/server/common/auth"
+	"go.temporal.io/server/common/config"
+)
+
+const defaultTimeout = time.Second * 10
+const rdsCaUrl = "https://s3.amazonaws.com/rds-downloads/rds-combined-ca-bundle.pem"
+
+var rdsAuthFn = AWSAuth.BuildAuthToken
+var rdsPemBundle string
+
+func init() {
+	RegisterPlugin("rds-iam-auth", NewRDSAuthPlugin(nil))
+}
+
+func fetchRdsCA(ctx context.Context) (string, error) {
+	_, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	client := http.DefaultClient
+	resp, err := client.Get(rdsCaUrl)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+	pem, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return base64.StdEncoding.EncodeToString(pem), nil
+}
+
+type RDSAuthPlugin struct {
+	awsConfig *aws.Config
+}
+
+func NewRDSAuthPlugin(awsConfig *aws.Config) AuthPlugin {
+	return &RDSAuthPlugin{
+		awsConfig: awsConfig,
+	}
+}
+
+func (plugin *RDSAuthPlugin) getToken(ctx context.Context, addr string, region string, user string, credentials aws.CredentialsProvider) (string, error) {
+	reqCtx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	return rdsAuthFn(reqCtx, addr, region, user, credentials)
+}
+
+func (plugin *RDSAuthPlugin) resolveAwsConfig(ctx context.Context) (*aws.Config, error) {
+	if plugin.awsConfig != nil {
+		return plugin.awsConfig, nil
+	}
+
+	reqCtx, cancel := context.WithTimeout(ctx, defaultTimeout)
+	defer cancel()
+
+	cfg, err := AWSConfig.LoadDefaultConfig(reqCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cfg, nil
+}
+
+func (plugin *RDSAuthPlugin) GetConfig(ctx context.Context, cfg *config.SQL) (*config.SQL, error) {
+	awsCfg, err := plugin.resolveAwsConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	token, err := plugin.getToken(ctx, cfg.ConnectAddr, awsCfg.Region, cfg.User, awsCfg.Credentials)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg.Password = token
+	cfg.ConnectProtocol = "tcp"
+
+	if cfg.ConnectAttributes == nil {
+		cfg.ConnectAttributes = map[string]string{}
+	}
+
+	// mysql requires this plugin to use the token as a password
+	if cfg.PluginName == "mysql" {
+		cfg.ConnectAttributes["allowCleartextPasswords"] = "true"
+	}
+
+	// if TLS is not configured, we default to the RDS CA
+	// this is required for mysql to send cleartext passwords
+	if cfg.TLS == nil {
+		if rdsPemBundle == "" {
```

Comment:
We should avoid using global variable. This can be a field of RDSAuthPlugin, and we can use sync.Once to run it once without having to worry about locking and race.

Reply:
✅ I followed the convention for sync.Once that saw in the rest of the project.